runtime/src/execute_cost_table.rs

```diff
@@ -80,17 +92,34 @@ impl ExecuteCostTable {
         self.table.get(key)
     }
 
+    pub fn initialize(&mut self, key: &Pubkey, cost: &u64) {
+        self.table.insert(*key, *cost);
+    }
+
     /// update-or-insert should be infallible. Query the result of upsert,
     /// often requires additional calculation, should be lazy.
-    pub fn upsert(&mut self, key: &Pubkey, value: u64) {
+    pub fn upsert(&mut self, key: &Pubkey, program_timing: &ProgramTiming) {
         let need_to_add = !self.table.contains_key(key);
         let current_size = self.get_count();
         if current_size == self.capacity && need_to_add {
             self.prune_to(&((current_size as f64 * PRUNE_RATIO) as usize));
         }
 
-        let program_cost = self.table.entry(*key).or_insert(value);
-        *program_cost = (*program_cost + value) / 2;
+        let datum = program_timing.accumulated_units / program_timing.count as u64;
```

Comment:
i don't find `datum` to be much more informative.  it's close to the `new_cost`, right?